File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/settings/MatchPlayClient.kt
package com.pillyliu.pinprofandroid.settings

import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

internal data class MatchPlayTournamentImportResult(
    val id: String,
    val name: String,
    val machineIds: List<String>,
)

internal object MatchPlayClient {
    fun fetchTournament(id: String): MatchPlayTournamentImportResult {
        val trimmed = id.trim()
        if (trimmed.isBlank()) error("Enter a valid tournament ID.")

        val root = JSONObject(fetchText("https://app.matchplay.events/api/tournaments/$trimmed?includeArenas=true"))
        val data = root.optJSONObject("data") ?: error("Tournament not found.")
        val tournamentId = data.optInt("tournamentId").takeIf { it > 0 }?.toString()
            ?: error("Tournament not found.")
        val name = data.optString("name").trim().ifBlank { "Tournament $tournamentId" }
        val arenas = data.optJSONArray("arenas") ?: JSONArray()
        val machineIds = LinkedHashSet<String>()
        for (i in 0 until arenas.length()) {
            val arena = arenas.optJSONObject(i) ?: continue
            val opdbId = arena.optString("opdbId").trim()
            if (opdbId.isNotBlank()) {
                machineIds += opdbId
            }
        }
        return MatchPlayTournamentImportResult(
            id = tournamentId,
            name = name,
            machineIds = machineIds.toList(),
        )
    }

    private fun fetchText(url: String): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = 15000
            readTimeout = 20000
            requestMethod = "GET"
        }
        val status = connection.responseCode
        if (status == 404) error("Tournament not found.")
        if (status !in 200..299) error("Match Play request failed ($status)")
        return connection.inputStream.bufferedReader().use { it.readText() }
    }
}

File: Pinball App 2/Pinball App 2/settings/MatchPlayClient.swift
import Foundation

struct MatchPlayTournamentImportResult: Hashable {
    let id: String
    let name: String
    let machineIDs: [String]
}

enum MatchPlayClient {
    private static let decoder = JSONDecoder()

    static func fetchTournament(id: String) async throws -> MatchPlayTournamentImportResult {
        let trimmed = id.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw MatchPlayClientError.invalidTournamentID
        }

        guard let url = URL(string: "https://app.matchplay.events/api/tournaments/\(trimmed)?includeArenas=true") else {
            throw MatchPlayClientError.invalidTournamentID
        }

        let (data, response) = try await URLSession.shared.data(from: url)
        if let httpResponse = response as? HTTPURLResponse,
           !(200 ... 299).contains(httpResponse.statusCode) {
            throw MatchPlayClientError.http(statusCode: httpResponse.statusCode)
        }

        let payload = try decoder.decode(MatchPlayTournamentResponse.self, from: data)
        let machineIDs = Array(
            NSOrderedSet(
                array: payload.data.arenas.compactMap { arena in
                    arena.opdbID?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank
                }
            )
        ).compactMap { $0 as? String }

        return MatchPlayTournamentImportResult(
            id: String(payload.data.tournamentID),
            name: payload.data.name,
            machineIDs: machineIDs
        )
    }
}

private enum MatchPlayClientError: LocalizedError {
    case invalidTournamentID
    case http(statusCode: Int)

    var errorDescription: String? {
        switch self {
        case .invalidTournamentID:
            return "Enter a valid tournament ID."
        case .http(let statusCode):
            if statusCode == 404 {
                return "Tournament not found."
            }
            return "Match Play request failed (\(statusCode))."
        }
    }
}

private struct MatchPlayTournamentResponse: Decodable {
    let data: MatchPlayTournamentData
}

private struct MatchPlayTournamentData: Decodable {
    let tournamentID: Int
    let name: String
    let arenas: [MatchPlayArena]

    enum CodingKeys: String, CodingKey {
        case tournamentID = "tournamentId"
        case name
        case arenas
    }
}

private struct MatchPlayArena: Decodable {
    let opdbID: String?

    enum CodingKeys: String, CodingKey {
        case opdbID = "opdbId"
    }
}

private extension String {
    var nilIfBlank: String? {
        isEmpty ? nil : self
    }
}
